keps/sig-api-machinery/2885-server-side-unknown-field-validation/README.md

```diff
@@ -248,26 +254,38 @@ implementation. What is the desired outcome and how do we measure success?.
 The "Design Details" section below is for the real
 nitty-gritty.
 -->
-We propose using an opt-in API mechanism (such as content-type header or query
-param) to indicate to the server that it should fail when the kubernetes object
-in the request body supplied to POST, PUT, and PATCH requests contains
-extra/unknown fields.
 
-Clients such as kubectl will continue to use the `--validate=true` flag as
-before, but instead of triggering validation on the client-side, it will
-instruct the server to validate for unknown fields server-side.
+We propose adding server-side schema validation, enabling the API server
+to validate that objects received in the request body in POST, PUT, and
+PATCH requests contain no unknown or duplicate fields.
 
-This change will be made in at least two steps, one where we introduce the
-server side validation and a second where we modify kubectl to use the
-server-side validation (and mark the existing client-side validation as
-deprecated).
+We introduce a query parameter that enables the user/client to indicate
+that server-side validation should either not be performed at all, provide
+warnings for the presence of unknown/duplicate fields, or error out in the
+presence of unknown/duplicate fields.
+
+For beta, we propose modifying the kubectl `--validate` flag to enable users
+to choose between client-side validation or server-side
+validation (either strict or warning)
 
 ### Notes/Constraints/Caveats (Optional)
 
 #### Future Work
-After server side unknown field validation is implemented we can begin work to
-deprecate and remove client side validate and have kubectl use server side
-validation instead. A separate KEP will be published for that.
+
+Upon successfully providing kubectl access to server-side validation via the
+`--validate` flag, an open question will remain as to the future of client
+side validation.
+
+While we would like kubectl to use server-side validation by default once it
+is GA, the question remains whether client-side validation be optionally
+available to kubectl users or completely deprecated.
+
+Maintaining client-side validation indefinitely provides the benefit of
+allowing users to validate their objects without needing access to a server.
+On the other hand, the downside of continuing to maintain client-side
+validation is that it uses openapiv2 which will be less actively supported
+as openapiv3 becomes standard and will cause client-side validation to
+continue to diverge from server-side validation.
```

Comment:
So what's the conclusion?
[Personally, I would like client-side validation to be deprecated and eventually completely replaced with server-side, but no matter what we should probably have some conclusion too what is our thinkng.]

Reply:
I suppose that full removal would need its own KEP.

Reply:
Sure - but let's say where we would like to head eventually.

Reply:
Added note about:
1. Separate KEP for CS validation deprecation
2. Alternate CS validation will be a pre-req for deprecation
3. kubeval is a potential candidate for that alternate validation.

Let me know if this is sufficient or if you think we need to do more to decide on what our future solution for CS validation will be now before we shoot for beta.

Reply:
Commented in a few places, but I don't think we should add mechanisms to explicitly request client-side validation.

I would expect the `--validate` flag to be intent-based, to use server-side validation when available, and N releases after server-side field validation hits GA, to drop support for client-side validation.

Reply:
> I would expect the --validate flag to be intent-based, to use server-side validation when available, and N releases after server-side field validation hits GA, to drop support for client-side validation.

+1 [that frames nicely what I was trying to say in my initial comment]

Reply:
Agreed. The kep now uses this language and I added a comment to explicitly spell out the "intent-based" nature of the `--validate` flag

Reply:
I'm not usually a big fan of flags whose behavior depends on some hard-to-assess logic (in this case, looking at the existence of a query param defined in the openapi). I'd rather want the behavior to be strictly defined and unconditional, with a proper error when used incorrectly (rather than the query param being merely ignored).